dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/reservedvalue/RandomGeneratorService.java

```diff
@@ -0,0 +1,207 @@
+ *         numbers and random UUID for alphanumerics.
+ *
+ *         x = lower case
+ *
+ *         X = upper case
+ *
+ *         # = digit
+ *
+ *         * = digit or lower case or upper case
+ *
+ *         For alphanumeric we have some corner cases:
+ *
+ *         - '*' it shuffles between lower and upper case and gets a number when
+ *         if finds it in UUID random string
+ *
+ *         - 'x' or 'X' if we find a number in UUID random string it gets the
+ *         alphabetic char at that position in order to maintain random pattern
+ *
+ */
+@Service
+@Setter( AccessLevel.PROTECTED )
+public class RandomGeneratorService implements Callable<List<String>>
+{
+
+    private final List<Character> lowercase = IntStream.range( 0, 26 ).mapToObj( n -> (char) (n + 'a') )
+        .collect( Collectors.toList() );
+
+    private final List<Character> uppercase = IntStream.range( 0, 26 ).mapToObj( n -> (char) (n + 'A') )
+        .collect( Collectors.toList() );
+
+    private String textPattern;
+
+    @Override
+    public List<String> call()
+        throws Exception
+    {
+        LinkedList<String> patterns = new LinkedList<>();
+
+        List<String> randomList = new ArrayList<>();
+
+        Pattern randomPattern = Pattern.compile( "[X]+|[x]+|[#]+|[*]+" );
+        Matcher matcher = randomPattern.matcher( textPattern );
+
+        while ( matcher.find() )
+        {
+            patterns.add( textPattern.substring( matcher.start(), matcher.end() ) );
+        }
+
+        for ( int j = 0; j < RANDOM_GENERATION_CHUNK; j++ )
+        {
+            StringBuilder stringBuilder = new StringBuilder();
+
+            for ( String pattern : patterns )
+            {
+                switch ( pattern.charAt( 0 ) )
+                {
+                case '*':
+                    setForRandomAll( stringBuilder, pattern );
+                    break;
+                case '#':
+                    stringBuilder.append( new BigInteger( 256, new SecureRandom() ).abs().toString(), 0,
+                        pattern.length() );
+                    break;
+                case 'X':
+                    setForRandomUpperCase( stringBuilder, pattern );
+                    break;
+                case 'x':
+                    setForRandomLowerCase( stringBuilder, pattern );
+                    break;
+                default:
+                    break;
+                }
+            }
+
+            randomList.add( stringBuilder.toString() );
+        }
+
+        return randomList;
+    }
+
+    private void setForRandomLowerCase( StringBuilder stringBuilder, String pattern )
+    {
+        int i = 0;
+        String randomUUIDForLower = UUID.randomUUID().toString();
+
+        while ( i < pattern.length() )
+        {
+            if ( Character.isLetter( randomUUIDForLower.charAt( i ) ) )
+            {
+                stringBuilder.append( randomUUIDForLower.charAt( i ) );
```

Comment:
what happen if one day pattern will be longer than UUID ? .chatAt(i) might overflow.